raft/src/main/java/org/apache/kafka/raft/internals/BatchBuilder.java

```diff
@@ -307,4 +309,38 @@ public int writeRecord(
         recordOutput.writeVarint(0);
         return ByteUtils.sizeOfVarint(sizeInBytes) + sizeInBytes;
     }
+
+    private int batchHeaderSizeInBytes() {
+        return AbstractRecords.recordBatchHeaderSizeInBytes(
+            RecordBatch.MAGIC_VALUE_V2,
+            compressionType
+        );
+    }
+
+    private int bytesNeededForRecords(
+        Collection<T> records,
+        ObjectSerializationCache serializationCache
+    ) {
+        long expectedNextOffset = nextOffset;
+        int bytesNeeded = 0;
+        for (T record : records) {
+            if (expectedNextOffset - baseOffset >= Integer.MAX_VALUE) {
+                return Integer.MAX_VALUE;
```

Comment:
Hmm.. If we return `Integer.MAX_VALUE`, then that will prevent us from being able to split when the number of records exceeds `Integer.MAX_VALUE`. Similarly, it might be worth hardening this logic by ensuring `bytesNeeded` does not overflow.  These cases are unlikely in practice, so we can save them for a follow-up if there is no clear way to handle them right now.

Reply:
True. I think the cleanest solution is to throw in this case.

I was thinking of returning `OptionalInt.of(Integer.MAX_VALUE)` and checking for it in the `BatchAccumulator`. With this solution we can't distinguish between the current batch plus the new records exceed this constraint. And the new records that are being added to an empty batch exceed this constraint.

I think throwing an exception is fine since this is a fatal error. This mean that we have at least `Integer.MAX_VALUE` which would most like result in a OOM exception somewhere else in the system.

Reply:
Since we are handling this case by raising an exception, is it worth checking for overflow of `bytesNeeded` as well?

Reply:
Updated it to use `Math.addExact`.